File: inference/app/logging.py
from __future__ import annotations

import json
import sys
from datetime import datetime, timezone
from typing import Any, Dict


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def log_json(level: str, message: str, service: str, version: str, trace_id: str) -> None:
    record: Dict[str, Any] = {
        "timestamp": _now_iso(),
        "level": level,
        "service": service,
        "version": version,
        "trace_id": trace_id,
        "message": message,
    }

    sys.stdout.write(json.dumps(record, ensure_ascii=False) + "\n")

File: ingestion/src/logging.ts
export class Logger {
  private service: string;
  private version: string;

  constructor(service: string, version: string) {
    this.service = service;
    this.version = version;
  }

  info(message: string, traceId: string): void {
    this.write("INFO", message, traceId);
  }

  warn(message: string, traceId: string): void {
    this.write("WARN", message, traceId);
  }

  error(message: string, traceId: string): void {
    this.write("ERROR", message, traceId);
  }

  private write(level: "INFO" | "WARN" | "ERROR", message: string, traceId: string): void {
    const record = {
      timestamp: new Date().toISOString(),
      level,
      service: this.service,
      version: this.version,
      trace_id: traceId,
      message
    };

    process.stdout.write(`${JSON.stringify(record)}\n`);
  }
}
